### DISCUSSION TITLE
Different recognition results on different GPU

### DISCUSSION THREAD

**Discussion by RudikAleksandr** [ultralytics/yolov5#9288] [Q&A] [answered]
I did a training for the same images with the same parameters for image size, number of epochs on different devices, but the recognition result for different devices is different. Why is this happening?

**First device configuration:**

- CPU: Intel Core i5-8400 2.80GHz
- GPU: GeForce GTX 1080/PCIe/SSE2
- OS: Ubuntu

**Recognition result after training:**
![first device result](https://user-images.githubusercontent.com/8377560/188420228-cadb7d04-4a86-420f-adb8-fbb421f322eb.jpg)

**Second device configuration:**

- CPU: Intel Core i5-8250U 1.60GHz 
- GPU: GeForce MX150
- OS: Windows 10

**Recognition result after training:**
![second device result](https://user-images.githubusercontent.com/8377560/188420436-4eb351d3-b4ed-45ad-a62c-aa48dc159676.jpg)

**Comment by glenn-jocher** [comment, marked as answer]
👋 Hello! Thanks for asking about **training reproducibility**. YOLOv5 🚀 uses a single training seed which is set here using the `init_seeds()` function. CPU and Single-GPU trainings should be fully reproducible with `torch>=1.12.0`. Multi-GPU DDP trainings are still not reproducible unfortunately. This is an open issue for us and we could use any help in tracking down this problem.
https://github.com/ultralytics/yolov5/blob/7215a0fb41a90d8a0bf259fa708dff608a1f0262/train.py#L104

This function sets `python`, `numpy` and `torch` seeds and updates `cudnn` settings:
https://github.com/ultralytics/yolov5/blob/7215a0fb41a90d8a0bf259fa708dff608a1f0262/utils/general.py#L198-L214

To set a new training seed for example:
```bash
python train.py --seed 3  # default seed=0
```

Note that even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!

**Reply by RudikAleksandr** [reply]
Hello, @glenn-jocher! 

Thanks for your detailed answer.

Could you also explain how to train the model on a local device so that the result tends to be the same for different developers with the same input? 

It is also necessary to understand that the model will recognize things in production, as we expect this on a local device.

**Reply by glenn-jocher** [reply]
@RudikAleksandr in general any differences should be relatively small. For torch 1.12 trainings should be fully reproducible on the same hardware. Even when not fully reproducible, i.e. torch 1.11 trainings should still be similar. On COCO for example the mAP std between trainings might be 1% of the total, or about 0.005 mAP.

Deployment should produce identical metrics on different hardware though no matter where they were trained.
